core/vm/contracts.go

```diff
@@ -1046,3 +1067,55 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
 	// Encode the G2 point to 256 bytes
 	return g.EncodePoint(r), nil
 }
+
+// SHA3-256 FIPS 202 standard implementation.
+type sha3fips struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *sha3fips) RequiredGas(input []byte) uint64 {
+	return uint64(len(input)+31)/32*params.Sha3FipsWordGas + params.Sha3FipsBaseGas
+}
+func (c *sha3fips) Run(input []byte) ([]byte, error) {
+	h := sha3.Sum256(input)
+	return h[:], nil
+}
+
+// Uncompressed Public Key recovery implementation.
+type ecrecoverPublicKey struct{}
+
+func (c *ecrecoverPublicKey) RequiredGas(input []byte) uint64 {
+	return params.EcrecoverGas
+}
+
+func (c *ecrecoverPublicKey) Run(input []byte) ([]byte, error) {
+	const ecrecoverPublicKeyInputLength = 128
+
+	input = common.RightPadBytes(input, ecrecoverPublicKeyInputLength)
+	// "input" is (hash, v, r, s), each 32 bytes
+	// but for ecrecover we want (r, s, v)
+
+	r := new(big.Int).SetBytes(input[64:96])
+	s := new(big.Int).SetBytes(input[96:128])
+	v := input[63]
```

Comment:
for ecrecover, it is `v := input[63] - 27`, what is the difference?

Reply:
v is the last byte of the signature, and is either 27 (0x1b) or 28 (0x1c). 
this is used to recover the points on elliptic curve.

```
In most implementations, the v is just 0 or 1 internally, but 27 was added as arbitrary number for signing Bitcoin messages and Ethereum adapted that as well.
```